### PR TITLE
feat(client): group by having types

### LINKED ISSUE #4872: Result type is not an array
## Problem

Getting a typecheck error on the result when accessing an element within the list

If I do:

```ts
const results = await client.agent.groupBy({
  by: ['location'],
  min: {
    commission: true
  },
})

console.log(results)

[
  { min: { commission: 0.14 } },
  { min: { commission: 0.13 } },
  { min: { commission: 0.14 } },
  { min: { commission: 0.11 } },
  { min: { commission: 0.13 } },
  { min: { commission: 0.11 } },
  { min: { commission: 0.12 } },
  { min: { commission: 0.12 } },
  { min: { commission: 0.15 } }
]
```

I get back a list. But if I try:

```ts
const results = await client.agent.groupBy({
  by: ['location'],
  min: {
    commission: true
  },
})

console.log(results[0])

Element implicitly has an 'any' type because expression of type '0' can't be used to index type 'Pick<AgentGroupByOutputType, "location"> & { min: { commission: number; }; }'.
  Property '0' does not exist on type 'Pick<AgentGroupByOutputType, "location"> & { min: { commission: number; }; }'.
```

## More details

I'm trying to test this query

```sql
-- Find the areas with the min commission
SELECT location, MIN(commission)
FROM agents
GROUP BY location;
```

**schema.prisma**

```schema
datasource db {
  provider = "sqlite"
  url      = "file:dev.db"
  default  = true
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["groupBy"]
}

model Agent {
  id         String @id
  name       String
  location   String
  commission Float
}
```

**index.ts**

```
import { PrismaClient } from '@prisma/client'

const rows: Array<[string, string, string, number]> = [
  ['A007', "Ramasundar", "Bangalore", 0.15],
  ['A003', "Alex", "London", 0.13],
  ['A008', "Alford", "New York", 0.12],
  ['A011', "Ravi Kumar", "Bangalore", 0.15],
  ['A010', "Santakumar", "Chennai", 0.14],
  ['A012', "Lucida", "San Jose", 0.12],
  ['A005', "Anderson", "Brisban", 0.13],
  ['A001', "Subbarao", "Bangalore", 0.14],
  ['A002', "Mukesh", "Mumbai", 0.11],
  ['A006', "McDen", "London", 0.15],
  ['A004', "Ivan", "Torento", 0.15],
  ['A009', "Benjamin", "Hampshair", 0.11],
]
const client = new PrismaClient()


async function main() {
  await client.$connect()

  // seed
  for (let row of rows) {
    await client.agent.upsert({
      where: {
        id: row[0],
      },
      create: {
        id: row[0],
        name: row[1],
        location: row[2],
        commission: row[3],
      },
      update: {},
    })
  }

  // query
  const results = await client.agent.groupBy({
    by: ['location'],
    select: {
      location: true
    },
    min: {
      commission: true
    },
  })

  console.log(results)
}

main().catch(console.error).finally(() => client.$disconnect())
```

## Details

```
@prisma/cli          : 2.14.0-dev.64
@prisma/client       : 2.14.0-dev.64
Current platform     : darwin
Query Engine         : query-engine 5d491261d382a2a5ffdc71de17072b0e409f1cc1 (at node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli 5d491261d382a2a5ffdc71de17072b0e409f1cc1 (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core 5d491261d382a2a5ffdc71de17072b0e409f1cc1 (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt 5d491261d382a2a5ffdc71de17072b0e409f1cc1 (at node_modules/@prisma/engines/prisma-fmt-darwin)
Studio               : 0.332.0
Preview Features     : groupBy
```

### PR DESCRIPTION
**Pull request by timsuchanek**
Closes https://github.com/prisma/prisma/issues/4872

### CODE REVIEW

File: src/packages/client/fixtures/blog/main.ts [hunk 1]
```diff
@@ -14,13 +14,65 @@ async function main() {
...
+  type C = 1
+  type D = ''
+
+  type IsObject<T extends any> = T extends Array<any>
+    ? 0
+    : T extends object
+    ? 1
+    : 0
+
+  type Ya = IsObject<D>
```
Comment (williamluke4) [inline src/packages/client/fixtures/blog/main.ts:72]: Hmmm

File: src/packages/client/package.json [hunk 2]
```diff
@@ -113,6 +113,7 @@
     "strip-indent": "3.0.0",
     "ts-jest": "26.4.4",
     "ts-node": "9.1.1",
+    "ts-toolbelt": "^8.0.7",
```
Comment (williamluke4) [inline src/packages/client/package.json:116]: It seems like you just extracted the utilities you wanted from ts-toolbelt. Do we actually use this anywhere?

**Comment by timsuchanek** [comment]
Thanks! Will merge already, as it's urgent and then fix in separate PR

**Comment by timsuchanek** [comment]
https://github.com/prisma/prisma/pull/4874

**Pull request merged**